services/imagery/src/backends/camera-base-backend.js

```diff
@@ -0,0 +1,167 @@
+import request from 'superagent';
+import addProtobuf from 'superagent-protobuf';
+
+import logger from '../common/logger';
+import { imagery, telemetry } from '../messages';
+
+import { removeOrientation, wait } from '../util';
+
+addProtobuf(request);
+
+export default class CameraBaseBackend {
+  /**
+   * Base which takes care of taking images on an interval.
+   *
+   * Takes care of telemetry as well. Requires extension to implement
+   * camera capture behavior.
+   *
+   * Extended classes may implement the acquire() and release(camera)
+   * functions, but must implmenent capture(camera).
+   */
+
+  /** Create a new camera base backend. */
+  constructor(imageStore, interval, telemetryUrl) {
+    this._imageStore = imageStore;
+    this._interval = interval;
+    this._telemetryUrl = telemetryUrl;
+    this._active = false;
+    this._runningMeta = 0;  // Number of running meta tasks.
+    this._endMeta = () => {};  // Used for stop().
+  }
+
+  /** Get the camera and then start the loop in the background. */
+  async start() {
+    this._active = true;
+
+    // Run an acquire function if created on implementor class.
+    if (this.acquire) {
+      this._camera = await this.acquire();
+    } else {
+      this._camera = { };
+    }
+
+    this._loop = this._runLoop();
+  }
+
+  /** Set a flag to kill the camera loop. */
+  async stop() {
+    this._active = false;
+
+    // Wait until the running loop has finished.
+    await this._loop;
+
+    // Wait until all the background tasks have finished.
+    await new Promise((resolve) => {
+      this._endMeta = () => this._runningMeta === 0 ? resolve() : null;
+      this._endMeta();  // In case it's already zero.
+    });
```

Comment:
Seems like you kind of reinvented `async.queue`, just with unlimited concurrency.

Reply:
Huh, yeah it seems like it.